File: integrations/hermes/web-searchhub/provider.py
"""SearchHub aggregator backend for hermes-agent.

Thin passthrough client for a self-hosted SearchHub instance
(REST endpoints /v1/search and /v1/extract).
"""

from agent.web_search_provider import WebSearchProvider, get_provider_env


class SearchHubProvider(WebSearchProvider):
    name = "searchhub"

    def __init__(self) -> None:
        self._base = (get_provider_env("SEARCHHUB_URL") or "").rstrip("/")
        self._token = get_provider_env("SEARCHHUB_TOKEN") or ""

    @property
    def display_name(self) -> str:
        return "SearchHub (self-hosted aggregator)"

    def is_available(self) -> bool:
        return bool(self._base and self._token)

    def supports_search(self) -> bool:
        return True

    def supports_extract(self) -> bool:
        return True

    def _headers(self) -> dict:
        return {"Authorization": f"Bearer {self._token}"}

    def _call(self, method: str, path: str, params: dict | None = None,
              json_body: dict | None = None, timeout: float = 30.0) -> dict:
        import httpx

        url = f"{self._base}{path}"
        try:
            if method == "GET":
                resp = httpx.get(url, params=params, headers=self._headers(), timeout=timeout)
            else:
                resp = httpx.post(url, json=json_body or {}, headers=self._headers(), timeout=timeout)
            body = resp.json()
        except Exception as e:
            return {"success": False, "error": str(e)}
        if resp.status_code != 200 or not isinstance(body, dict):
            return {"success": False, "error": body.get("error") if isinstance(body, dict) else f"http {resp.status_code}"}
        return body

    def search(self, query: str, limit: int = 5) -> dict:
        return self._call("GET", "/v1/search", params={"q": query, "limit": limit})

    def extract(self, urls, **kwargs) -> list:
        # Hermes 契约：extract 必须返回"每条 URL 一个 dict"的列表（见
        # agent/web_search_provider.py 的 ABC docstring），不是
        # {success, data} 信封——信封会被当成列表迭代，元素是字符串键，
        # 逐项 .get() 即抛 'str' object has no attribute 'get'。
        body = {"urls": list(urls)}
        for key in ("format", "include_raw", "max_chars"):
            if key in kwargs and kwargs[key] is not None:
                body[key] = kwargs[key]
        resp = self._call("POST", "/v1/extract", json_body=body, timeout=90.0)
        if not isinstance(resp, dict) or not resp.get("success"):
            error = resp.get("error", "extract failed") if isinstance(resp, dict) else str(resp)
            return [{"url": u, "title": "", "content": "", "raw_content": "",
                     "metadata": {}, "error": error} for u in urls]
        items = resp.get("data") or []
        out = []
        for it in items:
            if not isinstance(it, dict):
                out.append({"url": "", "title": "", "content": "", "raw_content": "",
                            "metadata": {}, "error": "malformed extract item"})
                continue
            out.append({
                "url": it.get("url", ""),
                "title": it.get("title", ""),
                "content": it.get("content", ""),
                "raw_content": it.get("raw_content", ""),
                "metadata": it.get("metadata") or {},
                "error": it.get("error"),
            })
        return out


def register(ctx) -> None:
    ctx.register_web_search_provider(SearchHubProvider())

File: integrations/dsh/searchhub-web/src/provider.ts
import { WebError } from '@deepseek-ai/dsh-web'
import type {
  WebFetchProvider, WebFetchRequest, WebFetchResult,
  WebSearchProvider, WebSearchRequest, WebSearchResult, WebSearchSource,
} from '@deepseek-ai/dsh-web'

export interface SearchHubProviderOptions {
  /** SearchHub REST base URL (no trailing slash). */
  baseURL: string
  /** Literal token; wins over resolveToken when present. */
  token?: string
  /** Resolve the current token for one operation. */
  resolveToken?: () => Promise<string | undefined>
}

async function requireToken(options: SearchHubProviderOptions): Promise<string> {
  const literal = options.token
  if (literal !== undefined && literal.length > 0) return literal
  if (options.resolveToken !== undefined) {
    const resolved = await options.resolveToken()
    if (resolved !== undefined && resolved.length > 0) return resolved
  }
  throw new WebError('SearchHub token is not configured', 'WEB_PROVIDER_ERROR')
}

export function isAvailable(options: SearchHubProviderOptions): boolean {
  return URL.canParse(options.baseURL)
    && ((options.token?.length ?? 0) > 0 || options.resolveToken !== undefined)
}

function mapSource(item: { title?: string; url?: string; description?: string; published_at?: string | null }): WebSearchSource | undefined {
  if (!item.url || item.url.length === 0) return undefined
  return {
    url: item.url,
    ...(item.title && item.title.length > 0 ? { title: item.title } : {}),
    ...(item.description && item.description.length > 0 ? { snippet: item.description } : {}),
    ...(item.published_at && item.published_at.length > 0 ? { publishedAt: item.published_at } : {}),
  }
}

export class SearchHubSearchProvider implements WebSearchProvider {
  readonly id = 'searchhub'

  constructor(private readonly resolveOptions: () => SearchHubProviderOptions) {}

  available(): boolean {
    return isAvailable(this.resolveOptions())
  }

  async search(request: WebSearchRequest, signal?: AbortSignal): Promise<WebSearchResult> {
    const options = this.resolveOptions()
    const token = await requireToken(options)
    const limit = request.maxResults ?? 5
    const url = new URL('/v1/search', options.baseURL)
    url.searchParams.set('q', request.query)
    url.searchParams.set('limit', String(limit))
    let body: unknown
    let resp: Response
    try {
      resp = await fetch(url, {
        headers: { Authorization: `Bearer ${token}` },
        signal,
      })
      body = await resp.json()
    } catch (err) {
      if (err instanceof WebError) throw err
      throw new WebError(`SearchHub search failed: ${(err as Error).message ?? String(err)}`, 'WEB_PROVIDER_ERROR')
    }
    const payload = body as { success?: boolean; data?: { web?: Array<Record<string, unknown>> }; error?: string }
    if (!resp.ok || payload.success === false || !payload.data?.web) {
      throw new WebError(payload.error ?? `SearchHub search http ${resp.status}`, 'WEB_PROVIDER_ERROR')
    }
    const sources = payload.data.web
      .map((item) => mapSource(item as { title?: string; url?: string; description?: string; published_at?: string | null }))
      .filter((s): s is WebSearchSource => s !== undefined)
    return { sources, truncated: false }
  }
}

export class SearchHubFetchProvider implements WebFetchProvider {
  readonly id = 'searchhub'

  constructor(private readonly resolveOptions: () => SearchHubProviderOptions) {}

  available(): boolean {
    return isAvailable(this.resolveOptions())
  }

  async fetch(request: WebFetchRequest, signal?: AbortSignal): Promise<WebFetchResult> {
    const options = this.resolveOptions()
    const token = await requireToken(options)
    let resp: Response
    let body: unknown
    try {
      resp = await fetch(new URL('/v1/extract', options.baseURL), {
        method: 'POST',
        headers: { Authorization: `Bearer ${token}`, 'Content-Type': 'application/json' },
        body: JSON.stringify({ urls: [request.url], format: 'markdown' }),
        signal,
      })
      body = await resp.json()
    } catch (err) {
      if (err instanceof WebError) throw err
      throw new WebError(`SearchHub fetch failed: ${(err as Error).message ?? String(err)}`, 'WEB_PROVIDER_ERROR')
    }
    const payload = body as { success?: boolean; data?: Array<Record<string, unknown>>; error?: string }
    if (!resp.ok || payload.success === false) {
      throw new WebError(payload.error ?? `SearchHub extract http ${resp.status}`, 'WEB_PROVIDER_ERROR')
    }
    const item = (payload.data ?? []).find((it) => it.url === request.url)
    if (!item) {
      throw new WebError(`SearchHub extract returned no result for ${request.url}`, 'WEB_PROVIDER_ERROR')
    }
    if (item.error) {
      throw new WebError(String(item.error), 'WEB_PROVIDER_ERROR')
    }
    return {
      url: request.url,
      statusCode: 200,
      truncated: false,
      body: { kind: 'text', content: String(item.content ?? item.raw_content ?? '') },
    }
  }
}
